model/map/src/main/java/org/keycloak/models/map/userSession/MapUserSessionProvider.java

```diff
@@ -0,0 +1,799 @@
+
+        LOG.tracef("removeUserSession(%s, %s)%s", realm, session, getShortStackTrace());
+
+        userSessionTx.getUpdatedNotRemoved(mcb).forEach(userSession -> {
+            userSession.getAuthenticatedClientSessions().values().stream().forEach(clientSessionTx::delete);
+            userSessionTx.delete(userSession.getId());
+        });
+
+    }
+
+    @Override
+    public void removeUserSessions(RealmModel realm, UserModel user) {
+        ModelCriteriaBuilder<UserSessionModel> mcb = userSessionStore.createCriteriaBuilder()
+                .compare(UserSessionModel.SearchableFields.REALM_ID, ModelCriteriaBuilder.Operator.EQ, realm.getId())
+                .compare(UserSessionModel.SearchableFields.USER_ID, ModelCriteriaBuilder.Operator.EQ, user.getId());
+
+        LOG.tracef("removeUserSessions(%s, %s)%s", realm, user, getShortStackTrace());
+
+        userSessionTx.getUpdatedNotRemoved(mcb).forEach(userSession -> {
+            userSession.getAuthenticatedClientSessions().values().stream().forEach(clientSessionTx::delete);
+            userSessionTx.delete(userSession.getId());
+        });
+    }
+
+    @Override
+    public void removeAllExpired() {
+        session.realms().getRealmsStream().forEach(this::removeExpired);
+    }
+
+    @Override
+    public void removeExpired(RealmModel realm) {
+        int currentTime = Time.currentTime();
+        int expired = currentTime - realm.getSsoSessionMaxLifespan();
+        int expiredRefresh = currentTime - realm.getSsoSessionIdleTimeout() - SessionTimeoutHelper.PERIODIC_CLEANER_IDLE_TIMEOUT_WINDOW_SECONDS;
+        int expiredRememberMe = currentTime - (realm.getSsoSessionMaxLifespanRememberMe() > 0 ?
+                realm.getSsoSessionMaxLifespanRememberMe() : realm.getSsoSessionMaxLifespan());
+        int expiredRefreshRememberMe = currentTime - (realm.getSsoSessionIdleTimeoutRememberMe() > 0 ?
+                realm.getSsoSessionIdleTimeoutRememberMe() : realm.getSsoSessionIdleTimeout()) -
+                SessionTimeoutHelper.PERIODIC_CLEANER_IDLE_TIMEOUT_WINDOW_SECONDS;
+        int expiredOffline = currentTime - realm.getOfflineSessionIdleTimeout() - SessionTimeoutHelper.PERIODIC_CLEANER_IDLE_TIMEOUT_WINDOW_SECONDS;
+        int clientExpired = Math.min(expired, expiredRememberMe);
+
+        // remove expired user sessions and its client sessions
+        ModelCriteriaBuilder<UserSessionModel> mcb = realmAndOfflineCriteriaBuilder(realm, false)
+                .compare(UserSessionModel.SearchableFields.IS_EXPIRED, ModelCriteriaBuilder.Operator.EQ,
+                        expiredRememberMe, expiredRefreshRememberMe, expired, expiredRefresh);
+        userSessionTx.getUpdatedNotRemoved(mcb)
+            .peek(userEntity -> {
+                userEntity.getAuthenticatedClientSessions().values().forEach(clientSessionTx::delete);
+                userEntity.clearAuthenticatedClientSessions();
```

Comment:
Thanks, the method looks much better now. Just last thing, is this line necessary? Now it is removed directly in Map storage without a transaction. If there will be some other transaction that accesses this userEntity and its clientSessions, there may be some ConcurrentModificationException. Maybe this can be omitted completely as this will be removed by `.forEach(userSessionTx::delete)`. WDYT?

Reply:
Yes, I removed it as it seems unnecessary.